src/lib/work_partitioner/pending_zkapp_command.ml

```diff
@@ -0,0 +1,48 @@
+(* NOTE: this module is where the real optimization happen. One assumption we
+   have is the order of merging is irrelvant to the correctness of the final
+   proof. Hence we're only using a counter `merge_remaining` to track have we
+   reach the final proof
+*)
+
+open Core_kernel
+open Snark_work_lib
+
+type t =
+  { job : (Spec.Single.t, Id.Single.t) With_job_meta.t
+        (* the original work being splitted, should be identical to Work_selector.work *)
+  ; unscheduled_segments : Spec.Sub_zkapp.t Queue.t
+  ; pending_mergable_proofs : Ledger_proof.Cached.t Deque.t
+        (* we may need to insert proofs to merge back to the queue, hence a Deque *)
+  ; mutable elapsed : Time.Stable.Span.V1.t
+  ; mutable merge_remaining : int
+  }
+
+let generate_merge ~(t : t) () =
+  let try_take2 (q : 'a Deque.t) : ('a * 'a) option =
+    match Deque.dequeue_front q with
+    | None ->
+        None
+    | Some fst -> (
+        match Deque.dequeue_front q with
+        | Some snd ->
+            Some (fst, snd)
+        | None ->
+            Deque.enqueue_front q fst ; None )
+  in
+  let open Option.Let_syntax in
+  let%map proof1, proof2 = try_take2 t.pending_mergable_proofs in
+  Spec.Sub_zkapp.Merge { proof1; proof2 }
+
+let generate_segment ~(t : t) () =
+  let open Option.Let_syntax in
+  let%map segment = Queue.dequeue t.unscheduled_segments in
+  segment
+
+let generate_job_spec (t : t) : Spec.Sub_zkapp.t option =
+  List.find_map ~f:(fun f -> f ()) [ generate_merge ~t; generate_segment ~t ]
```

Comment:
I'd consider writing this explicitly via `match` rather than through `find_map` function. Amount of code won't change, but it will be clearer